### DISCUSSION TITLE
Integrating Form Validation

### DISCUSSION THREAD

**Discussion by manosbatsis** [arkivanov/MVIKotlin#53] [Q&A] [answered]
Hi @arkivanov, thank you for all your great work. Hope you don't mind entertaining these:

1) From a practical point of view, what's might be the best way to integrate a form validation library like [form-builder](https://github.com/jkuatdsc/form-builder) or [chassis](https://github.com/boguszpawlowski/Chassis)? For example the latter provides a `StateFlow` but overall it seems to have too much overlap with MVIKotlin Stores, Reducers etc.
2) From a theoretical perspective, what would be the "right" way to integrate, e.g. a custom Store or using the form validation at a composable-level only?

**Comment by arkivanov** [comment, marked as answer]
Thanks for the question!

1. Indeed, both libraries seem to overlap with MVIKotlin, they both hold and update the state. Moreover, `form-builder` is tightly coupled with Jetpack Compose. I don't see any reasonable way of using those libraries together with MVIKotlin.
2. The `Store` holds the state and manages its changes consistently. You can just have all fields in the state class, and either validate values while they are being changed, or later on submission.

Here is a possible State class:

```kotlin
data class Field<T>(val value: T, val isError: Boolean)

data class State(
    val email: Field<String>,
    val phoneNumber: Field<String>,
)
```

In order to keep the `Store` as simple possible, the validation logic can be extracted into separate `Validator` classes and provided into the `Store` via DI.

**Reply by arkivanov** [reply]
If you really want to use such a library, then `chassis` looks more doable. It should be possible to pass a previously created form into the `Store` via DI, subscribe to its updates in the `Bootstrapper` and reflect the form state changes in the `Store` state. The `Executor` would receive Intents and redirect changes to the form. Does it make sense?

**Reply by manosbatsis** [reply]
Yup, it makes perfect sense, in a "how didn't ti thought of that" way, but then again i don't think i would :-) Thank you so much for all your work and help!
